pyo3-build-config/src/impl_.rs

```diff
@@ -486,47 +549,58 @@ pub fn any_cross_compiling_env_vars_set() -> bool {
         || env::var_os("PYO3_CROSS_PYTHON_VERSION").is_some()
 }
 
-fn cross_compiling() -> Result<Option<CrossCompileConfig>> {
+fn cross_compiling_from_cargo_env() -> Result<Option<CrossCompileConfig>> {
+    let host = cargo_env_var("HOST").unwrap();
+    let target = cargo_env_var("TARGET").unwrap();
+
+    if host == target {
+        // Definitely not cross compiling if the host matches the target
+        return Ok(None);
+    }
+
+    if target == "i686-pc-windows-msvc" && host == "x86_64-pc-windows-msvc" {
+        // Not cross-compiling to compile for 32-bit Python from windows 64-bit
+        return Ok(None);
+    }
+
+    let target_arch = cargo_env_var("CARGO_CFG_TARGET_ARCH").unwrap();
+    let target_vendor = cargo_env_var("CARGO_CFG_TARGET_VENDOR").unwrap();
+    let target_os = cargo_env_var("CARGO_CFG_TARGET_OS").unwrap();
+
+    cross_compiling(&host, &target_vendor, &target_arch, &target_os)
```

Comment:
Do we know that `TARGET` and `CARGO_CFG_TARGET_*` vars will always match?

Reply:
I think they will, yes.